File: qqbot-gateway/src/test/java/com/mieai/qqbot/gateway/FakeGatewayTransport.java
package com.mieai.qqbot.gateway;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

final class FakeGatewayTransport implements GatewayTransport {
    private final List<URI> connectedUrls = new ArrayList<>();
    private final List<FakeConnection> connections = new ArrayList<>();
    private final List<Listener> listeners = new ArrayList<>();

    @Override
    public CompletionStage<GatewayConnection> connect(URI gatewayUrl, Listener listener) {
        connectedUrls.add(gatewayUrl);
        FakeConnection connection = new FakeConnection();
        connections.add(connection);
        listeners.add(listener);
        return CompletableFuture.completedFuture(connection);
    }

    void emit(String payload) {
        latestListener().onText(payload);
    }

    void closeFromServer(int statusCode) {
        latestListener().onClosed(statusCode, "test close");
    }

    void fail(Throwable cause) {
        latestListener().onFailure(cause);
    }

    List<URI> connectedUrls() {
        return List.copyOf(connectedUrls);
    }

    List<FakeConnection> connections() {
        return List.copyOf(connections);
    }

    FakeConnection latestConnection() {
        return connections.getLast();
    }

    private Listener latestListener() {
        return listeners.getLast();
    }

    static final class FakeConnection implements GatewayConnection {
        private final List<String> sentPayloads = new ArrayList<>();
        private int closeCount;

        @Override
        public CompletionStage<Void> sendText(String payload) {
            sentPayloads.add(payload);
            return CompletableFuture.completedFuture(null);
        }

        @Override
        public CompletionStage<Void> close() {
            closeCount++;
            return CompletableFuture.completedFuture(null);
        }

        List<String> sentPayloads() {
            return List.copyOf(sentPayloads);
        }

        String latestPayload() {
            return sentPayloads.getLast();
        }

        int closeCount() {
            return closeCount;
        }
    }
}

File: qqbot-gateway/src/test/kotlin/com/mieai/qqbot/gateway/FakeGatewayTransport.kt
package com.mieai.qqbot.gateway

import java.net.URI
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage

internal class FakeGatewayTransport : GatewayTransport {
    private val connectedUrlsValue = mutableListOf<URI>()
    private val connectionsValue = mutableListOf<FakeConnection>()
    private val listeners = mutableListOf<GatewayTransport.Listener>()

    override fun connect(gatewayUrl: URI, listener: GatewayTransport.Listener): CompletionStage<GatewayConnection> {
        connectedUrlsValue += gatewayUrl
        val connection = FakeConnection()
        connectionsValue += connection
        listeners += listener
        return CompletableFuture.completedFuture<GatewayConnection>(connection)
    }

    fun emit(payload: String) = listeners.last().onText(payload)

    fun closeFromServer(statusCode: Int) = listeners.last().onClosed(statusCode, "test close")

    fun fail(cause: Throwable) = listeners.last().onFailure(cause)

    fun connectedUrls(): List<URI> = connectedUrlsValue.toList()

    fun connections(): List<FakeConnection> = connectionsValue.toList()

    fun latestConnection(): FakeConnection = connectionsValue.last()

    internal class FakeConnection : GatewayConnection {
        private val sentPayloadsValue = mutableListOf<String>()
        private var closeCountValue = 0

        override fun sendText(payload: String): CompletionStage<Void> {
            sentPayloadsValue += payload
            return CompletableFuture.completedFuture(null)
        }

        override fun close(): CompletionStage<Void> {
            closeCountValue++
            return CompletableFuture.completedFuture(null)
        }

        fun sentPayloads(): List<String> = sentPayloadsValue.toList()

        fun latestPayload(): String = sentPayloadsValue.last()

        fun closeCount(): Int = closeCountValue
    }
}
